### ISSUE TITLE
[RFC] Instruction-coreset selection (MAGIC) as a post-generation data-filtration stage

### ISSUE THREAD

**Issue by smellslikeml** [remyxai/VQASynth#95]
Considering whether VQASynth should add a new optional pipeline stage for **post-generation coreset selection** — filtering a generated VQA dataset down to a high-signal subset using influence signals derived from a reference VLM's forward pass. The reference method is [MAGIC](https://arxiv.org/abs/2605.26004), but the proposal is for the *capability*, not specifically MAGIC; the implementation just uses MAGIC's signals as a concrete first cut.

This is a **new stage**, not a replacement for `TagFilter`. The two operate on different things and at different points in the pipeline (details below).

## Current pipeline

```
generation stages → TagFilter (per-image CLIP-based include/exclude tag classification)
                  → published dataset
```

`TagFilter` operates per-image: does this image contain the tag? Keep or drop. It does not look at corpus-level properties, doesn't use the generated VQA pair (only the image), and doesn't have a notion of "quality" beyond the tag match.

## Proposed pipeline

```
generation stages → TagFilter      (unchanged — per-image relevance gate)
                  → CoresetSelector (NEW — VLM-influence ranking over generated VQA)
                  → published dataset
```

The new `CoresetSelector` stage:

- Runs a forward pass over the generated VQA pairs using **SpaceQwen** (Remyx's own spatial-reasoning VLM) as the scorer
- For each (image, instruction, answer) triple, computes three signals:
  - **Multimodal Gain** — `CE(answer | text-only) − CE(answer | image+text)`; how much does the image help the scorer?
  - **Bridging Relevance** — answer-token attention sharpness over visual tokens; is the answer actually grounded in image regions?
  - **Skill-Neuron Signature** — top-k FFN-neuron indices at fixed layers; what kind of spatial-reasoning skill does this sample elicit?
- Bucket samples by signature, allocate a target budget across buckets, select the top-quality samples per bucket
- Output: a subset of the generated dataset, optionally with the per-sample signals as columns for downstream debugging

## Why SpaceQwen as the scorer (not the paper's TinyLLaVA)

The MAGIC paper uses TinyLLaVA-2B as scorer for LLaVA-1.5-7B/13B targets — a general VL model scoring general-VL coreset selection. For VQASynth's case, the signals must be **domain-relative**, not universal:

- **Multimodal Gain becomes spatially-meaningful.** A general VL scorer might register MG on samples that need the image for *object identification*. SpaceQwen registers MG when the *spatial reasoning* genuinely improves with the image — which is the filter VQASynth actually wants.
- **Bridging Relevance becomes spatial.** Attention sharpness over visual tokens in a spatial model tracks attending to objects-in-relation, not just object identification.
- **Skill-neuron buckets become spatial skills.** TinyLLaVA's signature buckets are "object ID / OCR / color reasoning" — wrong diversity axis for spatial VQA. SpaceQwen's neurons (if the spatial fine-tune did its job) should bucket by spatial-skill: distance estimation, occlusion, relative positioning, scale invariance. *That's* the diversity to preserve.
- **No cross-scale assumption.** The paper assumes neuron indices transfer from TinyLLaVA-2B scorer to LLaVA-7B/13B targets — asserted but unvalidated. SpaceQwen scorer for a SpaceQwen-family target stays in-family.

Reference model default: `remyxai/SpaceQwen3-VL-2B-Thinking` (configurable). The 2B size matches the paper's scorer-is-cheap-relative-to-target pattern.

## Why this might be worth adding

1. **Quality lever orthogonal to TagFilter.** TagFilter removes samples whose image lacks the target tag; CoresetSelector removes samples that are technically on-topic but where the answer doesn't actually depend on the image (low Multimodal Gain) or isn't grounded in any visual region (diffuse attention). Two different failure modes, two different filters.
2. **Bounded compute cost.** A forward-only pass — no backprop. The MAGIC paper reports ~10 GPU-hours on 4×A100 to score 665K samples with a 2B model. Linear in dataset size, fixed in model size.
3. **Configurable target budget.** Downstream consumers training on VQASynth output sometimes want "give me your top-N samples" — currently that's not a thing the pipeline can provide.
4. **The signals themselves are useful debug surface.** Even without selection, attaching Multimodal Gain / Bridging Relevance / Skill-Neuron Signature to each sample as columns would let users diagnose generation-stage quality ("all my samples have Multimodal Gain near zero — even SpaceQwen doesn't think the image matters here").
5. **Validation IS the contribution.** Using Remyx's own spatial VLM as the scorer is something only the Remyx team is positioned to validate — adapting MAGIC's general-VL pipeline to the spatial-reasoning domain is the differentiated technical work.

## Why this is NOT a TagFilter replacement

MAGIC is *instruction-tuning coreset selection* — it picks subsets to train a stronger downstream VLM with fewer examples. TagFilter is *per-image binary classification during data generation*. Different mechanism, different point in the pipeline, different output shape. Replacing TagFilter with this would be the wrong move — TagFilter answers "is this image relevant?" and CoresetSelector answers "given that the image is relevant, is this VQA pair high-signal?". Both are useful; neither subsumes the other.

## Implementation plan

New top-level module `vqasynth/coreset/` (~600 LOC):

```
vqasynth/coreset/
├── __init__.py
├── signals.py     # Multimodal Gain, Bridging Relevance, Skill-Neuron — ~60 LOC
├── probing.py     # Forward-pass runner with hooks for FFN + attention — ~150 LOC
├── selection.py   # Bucket-based budget allocation algorithm — ~150 LOC
├── normalize.py   # Robust z-score scaling — ~15 LOC
└── stage.py       # Pipeline-stage wrapper: HF dataset in, HF dataset out — ~80 LOC
```

Plus:

- `docker/coreset_stage/process_coreset.py` — Docker entry for the new stage (mirrors the existing per-stage dockers)
- `tests/test_coreset_*.py` — unit tests against hand-computed values on toy inputs (signals + selection algorithm are deterministic and trivially unit-testable)
- `setup.py` — no new top-level deps; reuses existing `transformers` + `torch` pin. SpaceQwen loads via stock HF.

Default configuration:

- Reference model: `remyxai/SpaceQwen3-VL-2B-Thinking` (configurable)
- Layers probed: **uniformly-spaced through the LLM backbone** — for a 28-layer Qwen3-VL backbone the default is `{7, 14, 21, 27}`. Configurable per-checkpoint; needs a one-off ablation when changing the scorer model.
- Top-k per layer: `[1, 1, 2, 3]` as starting point (paper default); should be ablated against `[2, 2, 3, 4]` and `[1, 1, 1, 2]` since Qwen FFN width and sparsity dynamics may differ from the paper's LLaMA-based reference.
- Budget: ρ=0.6 eligibility, η=2.0 shortlist, α=β=0.5, τ=0.2, γ=0.05 per-bucket cap (paper defaults — these are method-level, not scorer-specific)
- **Default off** — stage runs only when `VQASYNTH_CORESET=on` is set, so existing pipelines see no behavior change

## Migration touches

- New module `vqasynth/coreset/` (additive)
- New Docker stage `docker/coreset_stage/` (additive)
- New env-var gate `VQASYNTH_CORESET` (additive)
- No existing files modified
- No dependency bump
- No dataset schema change unless `--emit-signals` is passed, in which case three new optional columns appear (`mg_score`, `br_score`, `signature`)

## Validation protocol

The paper's Table 4 / Table 5 results aren't directly reproducible — different scorer, different data, different domain. The replacement validation is a **held-out training comparison** on the actual deployment context:

1. **Held-out fine-tune comparison.** Generate a VQASynth dataset of fixed size N. Run CoresetSelector to produce a 20% subset. Compare:
   - Train a downstream spatial VLM (e.g., LLaVA-1.5-7B + LoRA, or a fresh SpaceQwen base) on the **20% MAGIC-selected** subset
   - Train an identical recipe on a **random 20%** subset of the same generation
   - Train an identical recipe on the **full 100%** generation
   - Evaluate all three on a held-out spatial reasoning benchmark (CV-Bench, BLINK-spatial, or a VQASynth-style held-out test set)
   - **Target: MAGIC-20% ≥ random-20% by ≥3pp on the spatial-reasoning metric; MAGIC-20% ≥ 0.95× full-100%**
2. **Sanity inspection on the signals themselves.**
   - Low-MG samples (bottom 5%) should be visually inspectable as image-independent — the answer could be guessed from the question alone (hallucinations, prior-bias answers)
   - Low-BR samples should attend to wrong regions (background, edges) rather than the spatial objects-in-relation the question asks about
   - Same-bucket samples should look spatially-similar by eyeball (e.g., one bucket clusters "distance between two objects" questions, another clusters "occlusion ordering")

## Known caveats

1. **Circularity risk** (and why it's probably fine). SpaceQwen was likely trained on VQASynth-generated data, so scoring VQASynth-generated samples with SpaceQwen carries a circularity — the scorer may over-favor patterns it has already seen. Real risk, but probably not fatal:
   - The probing is forward-only — no new training signal added
   - Bias toward patterns SpaceQwen has seen is *aligned* with the deployment distribution (downstream consumers want VQASynth-style data, not OOD data)
   - A low-MG sample that SpaceQwen rejects is probably a sample where the answer genuinely doesn't depend on the image — a generation-stage failure mode the filter SHOULD catch
   - The held-out fine-tune validation will surface circularity if it's actually hurting — MAGIC-20% performance dropping below random-20% would be the signal
2. **The paper has minor reproducibility gaps.** Exact form of `Norm(·)` isn't specified (default: robust z-score). Text-only cross-entropy handling — does the model still see a dummy image token, or is the projection skipped — isn't documented (default: skip projection). Largest-remainder method for bucket budget redistribution (default: Hagenbach-Bischoff). Each is a half-day ablation to validate.
3. **Layer-index recalibration is required when changing the scorer.** The paper's `{8,12,16,20}` was for 32-layer LLaMA-based LLaVA. Qwen3-VL backbone has different layer count and possibly different layer dynamics. Default `{7,14,21,27}` (for 28-layer) is a sensible starting point but should be sensitivity-tested.
4. **GPU requirement.** Forward-pass scoring needs a GPU; large-scale runs benefit from multi-GPU. Not a blocker for users who train VLMs on VQASynth output (they already have GPUs), but raises the floor for VQASynth's own CI — the validation step is GPU-heavy and should not be part of standard CI.
5. **Compute cost grows linearly with dataset size.** For users generating millions of samples, the probing pass is a non-trivial fraction of total cost. Should be opt-in, not on by default.

## Open questions

1. **Does VQASynth want to own coreset selection at all?** The alternative is to stay out of this layer and let downstream consumers run MAGIC (or similar) on VQASynth's output as a separate step. Pro of owning it: tighter integration, shared dataset schema, easier debug, validation goes through Remyx's spatial models. Con: pulls VQASynth into VLM-influence-scoring territory it doesn't currently occupy.
2. **Which SpaceQwen variant by default?** Options: `SpaceQwen3-VL-2B-Thinking` (smallest, newest base), `SpaceQwen2.5-VL-3B-Instruct` (more mature), `SpaceThinker-Qwen2.5VL-3B` (thinking-tuned). The 2B is cheapest and matches the paper's small-scorer pattern, but the 3B variants may have richer skill-neuron signatures. Worth a one-off comparison run before locking in a default.
3. **Output contract.** When the stage runs, does it produce a filtered dataset OR the original dataset annotated with the three signal columns + a `selected` boolean? Annotation is more flexible (downstream can re-filter at different budgets) but bigger on disk. Filtering is simpler but commits to the budget at generation time.
4. **Signal-only mode.** Should we ship a `--signals-only` mode that computes the three signals and stops, without running the selection algorithm? Useful when users want their own selection logic but want VQASynth to do the expensive scoring pass once.
5. **Validation acceptance bar.** What's the minimum result we'd accept before merging? MAGIC-20% must beat random-20% by some margin (3pp? 5pp?) on a named benchmark? Or just "the math is right + signals pass eyeball inspection," ship and let usage tell us if the numbers hold up?
6. **Circularity check protocol.** If SpaceQwen-on-VQASynth shows a circularity bias, should the scorer be a *different* spatial VLM (e.g., off-the-shelf spatial-finetune from another team) to break the loop? Or do we accept the alignment-with-deployment-distribution as the right trade-off?

Looking for thoughts before scoping a PR.

**Comment by smellslikeml** [comment]
Leaving open. Post-generation coreset selection is a real roadmap item — TagFilter operates per-image while MAGIC-style influence signals operate at corpus level; they're complementary, not redundant. Keeping the RFC discoverable while other pipeline work has priority.
